### PR TITLE
Feature/issue 126 capture tfl disruption data

### LINKED ISSUE #126: Phase 0: Capture TfL Disruption Data
## Problem

Currently, `DisruptionResponse` schema lacks `affectedRoutes` field. TfL API provides detailed route and station data via `StatusByIdsByPathIdsQueryDetail(detail=True)`, but we discard it in `_extract_disruption_from_line_status()` (line 943-992).

## Proposed Solution

Extend schema to capture TfL's disruption detail data and store it for future use.

## Schema Changes

**File: `backend/app/schemas/tfl.py`**
```python
class AffectedRouteInfo(BaseModel):
    """TfL route variant affected by disruption."""
    name: str  # e.g., "Cockfosters → Heathrow T5"
    direction: str  # "inbound" or "outbound"
    affected_stations: list[str]  # NaPTAN codes from routeSectionNaptanEntrySequence

class DisruptionResponse(BaseModel):
    # ... existing fields ...
    affected_routes: list[AffectedRouteInfo] | None = None  # NEW
```

## API Changes

None (internal schema only).

## Data Import Logic Changes

**File: `backend/app/services/tfl_service.py:943-992`**
- Extract `disruption.affectedRoutes` from TfL API response
- Map `routeSectionNaptanEntrySequence` to `affected_stations`
- Handle cases where `affectedRoutes` is missing (set to None)

## Task List

- [ ] Add `AffectedRouteInfo` schema to `backend/app/schemas/tfl.py`
- [ ] Add `affected_routes` field to `DisruptionResponse` (nullable, default None)
- [ ] Update `_extract_disruption_from_line_status()` to extract affected routes data
- [ ] Handle missing/malformed `affectedRoutes` gracefully (return None)
- [ ] Write unit tests for schema serialization/deserialization
- [ ] Write unit tests for extraction logic with/without affected routes
- [ ] Write integration test with real TfL API response structure
- [ ] Verify 100% code coverage for new code

## ADR Updates

**File: `docs/adr/07-external-apis.md`**
- Add section documenting that we capture detailed disruption data from `StatusByIdsByPathIdsQueryDetail(detail=True)`

## Testing Requirements (100% Coverage)

- Unit: Schema validation (with/without affected_routes)
- Unit: Extraction with mock TfL responses (5+ test cases)
- Integration: End-to-end TfL fetch → schema mapping
- Edge cases: Empty routes, missing fields, null values

## Related Issues

Part of #125 (Route Station Inverted Index)

## Estimated Time

1-2 days

### PR DESCRIPTION
**Pull request by mnbf9rca**
resolves #126

## Summary by Sourcery

Include detailed affected route information (route segments and station lists) in disruption responses by introducing helper functions, updating the extraction logic, extending the schema, and adding comprehensive tests and documentation.

New Features:
- Capture affected routes with station sequences from TfL disruption responses

Enhancements:
- Add pure helper functions for parsing timestamps, extracting NaPTAN codes, and building affected route info
- Update disruption extraction to include affected routes, reason, and created timestamp using helper methods

Documentation:
- Add AffectedRouteInfo schema and update DisruptionResponse to include affected_routes
- Document the decision to capture affected routes in the ADR for external APIs

Tests:
- Add unit tests for all new helper functions and integration tests covering various affected routes scenarios

### CODE REVIEW

**Comment by mnbf9rca** [comment]
## PR Review Response

Thank you for the comprehensive review feedback\! Here's my approach to each suggestion:

### ✅ ACCEPTING (10 items)

**Code Quality Improvements (5):**
- Replace `if/else` with `or` operator (line 1100)
- Use ternary expression in `_extract_naptan_id_from_stop_point`
- Use walrus operators in 3 functions (lines 1011, 1116, 1140)

**Timestamp Handling Fix (1):**
- Change `_parse_tfl_timestamp` to return `datetime | None` instead of `datetime.now()` fallback
- Add warning logs for invalid timestamps
- Rationale: `DisruptionResponse.created_at` is already `datetime | None`, so returning `None` accurately represents "unknown creation time" vs. falsely claiming "created just now"

**Test Coverage (3):**
- Add test for StopPoint missing naptanId
- Add test for affected_stations=None
- Add test for sequence with missing data

**Test Refactoring (1):**
- Parametrize repetitive tests using `pytest.mark.parametrize`

---

### ❌ REJECTING (1 item)

**Magic Constant Extraction:**
- Not extracting `"0001-01-01T00:00:00"` to named constant
- **Rationale:** YAGNI principle - Pydantic models handle this appropriately. Will add constant later if it becomes a problem.

---

### 📋 DEFERRED (2 items)

**Architecture Refactoring:**
- Not extracting helpers to separate module
- Not splitting TfLService currently
- **Rationale:** Following KISS/YAGNI for hobby project. TfLService (~1200 lines) isn't large enough to warrant splitting yet. Helpers are tightly coupled to service.
- **Action:** Creating separate issue to track this for future consideration when service grows larger.

---

### Implementation Plan

1. Create issue for future TfLService refactoring
2. Apply all accepted code changes
3. Verify 100% test coverage maintained
4. Run pre-commit hooks
5. Update this PR with status

Will update inline comments as changes complete.

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -940,6 +947,208 @@ def deduplicate_stations_by_hub(self, stations: list[Station], line_filter: str
...
+
+        # Process all routes, filtering out None values
+        valid_routes = [
+            route_info
+            for route in tfl_affected_routes
+            if (route_info := TfLService._process_affected_route(route)) is not None
+        ]
+
+        # Return None if no valid routes were extracted
+        return valid_routes if valid_routes else None
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Implemented - changed to `return valid_routes or None`

File: backend/app/services/tfl_service.py [hunk 2]
```diff
@@ -940,6 +947,208 @@ def deduplicate_stations_by_hub(self, stations: list[Station], line_filter: str
...
+
+        Args:
+            stop_point: TfL API StopPoint object
+
+        Returns:
+            NaPTAN ID string, or None if not found
+        """
+        if not stop_point:
+            return None
+        return getattr(stop_point, "naptanId", None)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Implemented - changed to ternary expression

File: backend/app/services/tfl_service.py [hunk 3]
```diff
@@ -940,6 +947,208 @@ def deduplicate_stations_by_hub(self, stations: list[Station], line_filter: str
...
+            List of NaPTAN code strings (empty if sequence is None or contains no valid codes)
+        """
+        if not sequence:
+            return []
+
+        naptan_codes = []
+        for seq_item in sequence:
+            stop_point = getattr(seq_item, "stopPoint", None)
+            naptan_id = TfLService._extract_naptan_id_from_stop_point(stop_point)
+            if naptan_id:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Implemented - using walrus operator

File: backend/app/services/tfl_service.py [hunk 4]
```diff
@@ -940,6 +947,208 @@ def deduplicate_stations_by_hub(self, stations: list[Station], line_filter: str
...
+
+        Args:
+            line_status: TfL API LineStatus object
+            disruption: TfL API Disruption object (may be None)
+
+        Returns:
+            Reason string, or None if not found in either source
+        """
+        reason: str | None = getattr(line_status, "reason", None)
+        if reason:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Implemented - using walrus operator with explicit None check

File: backend/app/services/tfl_service.py [hunk 5]
```diff
@@ -940,6 +947,208 @@ def deduplicate_stations_by_hub(self, stations: list[Station], line_filter: str
...
+
+        Args:
+            line_status: TfL API LineStatus object
+            disruption: TfL API Disruption object (may be None)
+
+        Returns:
+            Timestamp string, or None if not found in either source
+        """
+        created_str: str | None = getattr(line_status, "created", None)
+        if created_str:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Implemented - using walrus operator with explicit None check

File: backend/tests/test_tfl_service.py [hunk 6]
```diff
@@ -1574,6 +1588,285 @@ async def test_fetch_stations_skip_validation_calls_api(
...
+    )
+
+    result = TfLService._extract_naptan_id_from_stop_point(stop_point)
+    assert result == "940GZZLUKSX"
+
+
+def test_extract_naptan_id_from_stop_point_none() -> None:
+    """Test extracting NaPTAN ID from None StopPoint."""
+    result = TfLService._extract_naptan_id_from_stop_point(None)
+    assert result is None
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_tfl_service.py:1630]: ✅ Implemented - added test `test_extract_naptan_id_from_stop_point_missing_naptan_id` and additional edge case tests

**Comment by mnbf9rca** [comment]
## ✅ All PR Review Suggestions Implemented

All accepted suggestions have been implemented, tested, and verified. Changes are ready for review.

---

### 🎯 Changes Summary

#### Code Quality Improvements (5 items)
- ✅ Line 1100: Changed to `return valid_routes or None`
- ✅ Lines 986-988: Converted to ternary expression in `_extract_naptan_id_from_stop_point`
- ✅ Line 1009: Added walrus operator in `_extract_naptan_codes_from_sequence`
- ✅ Line 1123: Added walrus operator in `_extract_reason_from_sources` (with explicit None check for mypy)
- ✅ Line 1146: Added walrus operator in `_extract_created_timestamp` (with explicit None check for mypy)

#### Timestamp Handling Fix (CRITICAL)
- ✅ Changed `_parse_tfl_timestamp` return type: `datetime` → `datetime | None`
- ✅ Return `None` instead of `datetime.now()` for invalid timestamps
- ✅ Added structured warning logs for:
  - None/empty timestamps
  - Sentinel value "0001-01-01T00:00:00"
  - Parse failures
- ✅ Updated docstring to reflect logging side effects

**Rationale**: `DisruptionResponse.created_at` is `datetime | None`, so returning `None` accurately represents "unknown creation time" vs. falsely claiming "created just now"

#### Test Coverage (3 new tests + parametrization)
- ✅ Added `test_extract_naptan_id_from_stop_point_missing_naptan_id`
- ✅ Added `test_extract_naptan_codes_from_sequence_missing_data`
- ✅ Added `test_build_affected_route_info_none_stations`
- ✅ Parametrized timestamp tests (4 cases → 1 parametrized test)
- ✅ Parametrized process_affected_route tests (3 cases → 1 parametrized test)
- ✅ Parametrized extract_affected_routes tests (2 cases → 1 parametrized test)

#### Deferred Work
- ✅ Created issue #132 to track TfLService refactoring for future consideration

---

### ✅ Verification

**Tests**: All 213 tests passing ✅  
**Coverage**: 
- `tfl_service.py`: 96.33% (exceeds 95% target) ✅
- `schemas/tfl.py`: 100% ✅

**Pre-commit hooks**: All passing ✅
- ruff (format + lint)
- mypy --strict
- prettier
- eslint
- tsc

**Inline comments**: All resolved ✅

---

### 📋 What Was Rejected

**Magic Constant**: Not extracting `"0001-01-01T00:00:00"` to named constant (YAGNI - Pydantic handles this)

---

Ready for final review and merge! 🚀

File: backend/app/services/tfl_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Already implemented in previous review cycle

File: backend/app/services/tfl_service.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Already implemented in previous review cycle

File: backend/app/services/tfl_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Already implemented in previous review cycle

File: backend/app/services/tfl_service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Already implemented in previous review cycle

File: backend/app/services/tfl_service.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Already implemented in previous review cycle

File: backend/tfl_responses/Line_RouteSequenceByPathIdPathDirectionQueryServiceTypesQueryExcludeCrowding.txt [hunk 7]
```diff
@@ -0,0 +1,4962 @@
...
+            "lat": 51.488337,
+            "lon": -0.105963
+        }, {
+            "$type": "Tfl.Api.Presentation.Entities.MatchedStop, Tfl.Api.Presentation.Entities",
+            "stationId": "940GZZNEUGST",
+            "icsId": "1002196",
+            "topMostParentId": "940GZZNEUGST",
+            "modes": ["tube"],
+            "stopType": "NaptanMetroStation",
+            "zone": "6",
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tfl_responses/Line_RouteSequenceByPathIdPathDirectionQueryServiceTypesQueryExcludeCrowding.txt:4875]: This is test fixture data from the actual TfL API response. The zone data comes directly from TfL's API and is used for testing our parsing logic. We're not modifying or correcting TfL's data - we're testing that we handle it correctly as received.
